In [0]:
#Import pandas libraries
import pandas as pd

#Import numpy libraries
import numpy as np

import seaborn as sns
import matplotlib.pyplot as plt
%matplotlib inline


#Import category encoders
import category_encoders as ce


#Import sklearn libraries
from sklearn.impute import SimpleImputer
from sklearn import metrics
#  import classifier libraries
from sklearn.ensemble import AdaBoostClassifier
from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import GradientBoostingClassifier
from sklearn.model_selection import train_test_split
from sklearn.ensemble import RandomForestClassifier
from sklearn.preprocessing import RobustScaler


# Import Suport Vector Classifier module from svm library. We'll use SVC to model our data
from sklearn.svm import SVC,LinearSVC
from sklearn.model_selection import train_test_split
# Import scikit-learn metrics module for accuracy calculation
from sklearn.metrics import  accuracy_score
from sklearn.metrics import confusion_matrix

#Import warnings filter libraries
import warnings
warnings.filterwarnings('ignore')

In [0]:
#Load the dataset
df = pd.read_csv('hypothyroid.csv')

# Check shape of the dataset
df.shape

(3163, 26)

In [0]:
#Check the first five rows of our dataset
df.head()

status age sex on_thyroxine  ... FTI_measured FTI TBG_measured  TBG
0  hypothyroid  72   M            f  ...            y  10            n  NaN
1  hypothyroid  15   F            t  ...            y  17            n  NaN
2  hypothyroid  24   M            f  ...            y   0            n  NaN
3  hypothyroid  24   F            f  ...            y   6            n  NaN
4  hypothyroid  77   M            f  ...            y  44            n  NaN

[5 rows x 26 columns]

In [0]:
#Check null values

df.isnull().sum()

status                       0
age                          0
sex                          0
on_thyroxine                 0
query_on_thyroxine           0
on_antithyroid_medication    0
thyroid_surgery              0
query_hypothyroid            0
query_hyperthyroid           0
pregnant                     0
sick                         0
tumor                        0
lithium                      0
goitre                       0
TSH_measured                 0
TSH                          0
T3_measured                  0
T3                           0
TT4_measured                 0
TT4                          0
T4U_measured                 0
T4U                          0
FTI_measured                 0
FTI                          0
TBG_measured                 0
TBG                          0
dtype: int64

Our dataset has no null values. However if we check on a single column there is a ? which should be converted to NAN value

### **Replace ? with NAN**

In [0]:
# Count unique elements in each column including NaN
uniqueValues = df.nunique(dropna=False)
uniqueValues


Count Unique values in each column including NaN
status                         2
age                           93
sex                            3
on_thyroxine                   2
query_on_thyroxine             2
on_antithyroid_medication      2
thyroid_surgery                2
query_hypothyroid              2
query_hyperthyroid             2
pregnant                       2
sick                           2
tumor                          2
lithium                        2
goitre                         2
TSH_measured                   2
TSH                          240
T3_measured                    2
T3                            70
TT4_measured                   2
TT4                          269
T4U_measured                   2
T4U                          159
FTI_measured                   2
FTI                          281
TBG_measured                   2
TBG                           53
dtype: int64




*   Column sex has 3 unique values, to check we print the unique values.



In [0]:
#For example the below column has a unique value ? which needs to be replaced  
df['sex'].unique()

array(['M', 'F', '?'], dtype=object)

In [0]:
#Replace all rows with ? to nan
df.replace('?',np.nan,inplace=True)
df.head()

status age sex on_thyroxine  ... FTI_measured FTI TBG_measured  TBG
0  hypothyroid  72   M            f  ...            y  10            n  NaN
1  hypothyroid  15   F            t  ...            y  17            n  NaN
2  hypothyroid  24   M            f  ...            y   0            n  NaN
3  hypothyroid  24   F            f  ...            y   6            n  NaN
4  hypothyroid  77   M            f  ...            y  44            n  NaN

[5 rows x 26 columns]

In [0]:
#Check for missing values
df.isnull().sum()

status                          0
age                           446
sex                            73
on_thyroxine                    0
query_on_thyroxine              0
on_antithyroid_medication       0
thyroid_surgery                 0
query_hypothyroid               0
query_hyperthyroid              0
pregnant                        0
sick                            0
tumor                           0
lithium                         0
goitre                          0
TSH_measured                    0
TSH                           468
T3_measured                     0
T3                            695
TT4_measured                    0
TT4                           249
T4U_measured                    0
T4U                           248
FTI_measured                    0
FTI                           247
TBG_measured                    0
TBG                          2903
dtype: int64

In [0]:
#Confirm that the ? value has been replaced.
df['sex'].unique()


array(['M', 'F', nan], dtype=object)

In [0]:
#Checking data types of our dataset

df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3163 entries, 0 to 3162
Data columns (total 26 columns):
status                       3163 non-null object
age                          2717 non-null object
sex                          3090 non-null object
on_thyroxine                 3163 non-null object
query_on_thyroxine           3163 non-null object
on_antithyroid_medication    3163 non-null object
thyroid_surgery              3163 non-null object
query_hypothyroid            3163 non-null object
query_hyperthyroid           3163 non-null object
pregnant                     3163 non-null object
sick                         3163 non-null object
tumor                        3163 non-null object
lithium                      3163 non-null object
goitre                       3163 non-null object
TSH_measured                 3163 non-null object
TSH                          2695 non-null object
T3_measured                  3163 non-null object
T3                           2468 non-null 

There are numerical variables that need to be converted to numeric type.

In [0]:
num = ['age','TSH','T3','TT4','T4U','FTI']
num

['age', 'TSH', 'T3', 'TT4', 'T4U', 'FTI']

In [0]:
categorical= ['status','sex', 'on_thyroxine', 'query_on_thyroxine','on_antithyroid_medication', 'thyroid_surgery', 'query_hypothyroid',
      'query_hyperthyroid', 'pregnant', 'sick', 'tumor', 'lithium', 'goitre','TSH_measured','T3_measured','TT4_measured',
      'T4U_measured','FTI_measured','TBG_measured', 'TBG']

In [0]:
#convert  object to numerical
df[num] = df[num].apply(pd.to_numeric)
df.dtypes

status                        object
age                          float64
sex                           object
on_thyroxine                  object
query_on_thyroxine            object
on_antithyroid_medication     object
thyroid_surgery               object
query_hypothyroid             object
query_hyperthyroid            object
pregnant                      object
sick                          object
tumor                         object
lithium                       object
goitre                        object
TSH_measured                  object
TSH                          float64
T3_measured                   object
T3                           float64
TT4_measured                  object
TT4                          float64
T4U_measured                  object
T4U                          float64
FTI_measured                  object
FTI                          float64
TBG_measured                  object
TBG                           object
dtype: object

In [0]:
# To confirm they have been converted,split numerical variables from categorical variables
numerical_variables = [col for col in df.columns if df[col].dtypes != 'O']
numerical_variables


['age', 'TSH', 'T3', 'TT4', 'T4U', 'FTI']

In [0]:
#Get all categorical variables
categorical_variables = [col for col in df.columns if df[col].dtypes == 'O']
categorical_variables

['status',
 'sex',
 'on_thyroxine',
 'query_on_thyroxine',
 'on_antithyroid_medication',
 'thyroid_surgery',
 'query_hypothyroid',
 'query_hyperthyroid',
 'pregnant',
 'sick',
 'tumor',
 'lithium',
 'goitre',
 'TSH_measured',
 'T3_measured',
 'TT4_measured',
 'T4U_measured',
 'FTI_measured',
 'TBG_measured',
 'TBG']

In [0]:
#Check for missing values
df.isnull().sum()

status                          0
age                           446
sex                            73
on_thyroxine                    0
query_on_thyroxine              0
on_antithyroid_medication       0
thyroid_surgery                 0
query_hypothyroid               0
query_hyperthyroid              0
pregnant                        0
sick                            0
tumor                           0
lithium                         0
goitre                          0
TSH_measured                    0
TSH                           468
T3_measured                     0
T3                            695
TT4_measured                    0
TT4                           249
T4U_measured                    0
T4U                           248
FTI_measured                    0
FTI                           247
TBG_measured                    0
TBG                          2903
dtype: int64

In [0]:
#Fill missing values of numerical variables


# Use simple imputer to fill missing values with the mean
impute = SimpleImputer(strategy ='mean')
df[numerical_variables] = impute.fit_transform(df[numerical_variables])

In [0]:
# Fill missing values for categorical data
df['sex'].fillna(df['sex'].mode()[0], inplace=True)
df['TBG'].fillna(df['TBG'].mode()[0], inplace=True) 

#Confirm there are no missing values
df.isnull().sum()   

status                       0
age                          0
sex                          0
on_thyroxine                 0
query_on_thyroxine           0
on_antithyroid_medication    0
thyroid_surgery              0
query_hypothyroid            0
query_hyperthyroid           0
pregnant                     0
sick                         0
tumor                        0
lithium                      0
goitre                       0
TSH_measured                 0
TSH                          0
T3_measured                  0
T3                           0
TT4_measured                 0
TT4                          0
T4U_measured                 0
T4U                          0
FTI_measured                 0
FTI                          0
TBG_measured                 0
TBG                          0
dtype: int64

##**Data Preprocessing**

In [0]:
#We define x and y
y = df['status']
y


0       hypothyroid
1       hypothyroid
2       hypothyroid
3       hypothyroid
4       hypothyroid
           ...     
3158       negative
3159       negative
3160       negative
3161       negative
3162       negative
Name: status, Length: 3163, dtype: object

In [0]:
#Change our target values(y) to a binary

y =df['status']= np.where(df['status']=='hypothyroid',0,1)
print(y)
df['status'].value_counts()


[0 0 0 ... 1 1 1]


1    3012
0     151
Name: status, dtype: int64



*   1 means it's negative
*   0 means it's hypothyroid



In [0]:
X = df.drop(['status'], axis=1)
X

age sex on_thyroxine  ...    FTI TBG_measured TBG
0     72.0   M            f  ...   10.0            n  28
1     15.0   F            t  ...   17.0            n  28
2     24.0   M            f  ...    0.0            n  28
3     24.0   F            f  ...    6.0            n  28
4     77.0   M            f  ...   44.0            n  28
...    ...  ..          ...  ...    ...          ...  ..
3158  58.0   F            f  ...   95.0            n  28
3159  29.0   F            f  ...   98.0            n  28
3160  77.0   M            f  ...  104.0            n  28
3161  74.0   F            f  ...  137.0            n  28
3162  56.0   F            t  ...  143.0            n  28

[3163 rows x 25 columns]

In [0]:
from sklearn.model_selection import train_test_split

#Split our dataset train dataset size is 80% test datset is 20%
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.2, random_state = 0)

print(X_train.shape, X_test.shape)

(2530, 25) (633, 25)


In [0]:
!pip install sklearn

In [0]:
!pip install category_encoders

     |████████████████████████████████| 102kB 2.2MB/s 


In [0]:

# encode categorical variables with one-hot encoding to numeric

encoder = ce.OneHotEncoder(cols=['sex','on_thyroxine','query_on_thyroxine','on_antithyroid_medication','thyroid_surgery','query_hypothyroid','query_hyperthyroid','pregnant','sick','tumor','lithium','goitre','TSH_measured','T3_measured','TT4_measured','T4U_measured','FTI_measured','TBG_measured','TBG'])

X_train = encoder.fit_transform(X_train)

X_test = encoder.transform(X_test)



In [0]:
print(X_train.head(4))

       age  sex_1  sex_2  on_thyroxine_1  ...  TBG_47  TBG_48  TBG_49  TBG_50
2949  70.0      1      0               1  ...       0       0       0       0
1763  87.0      1      0               0  ...       0       0       0       0
349   40.0      1      0               0  ...       0       0       0       0
988   74.0      1      0               0  ...       0       0       0       0

[4 rows x 91 columns]


In [0]:
#Confirm there is no nan in train dataset.
np.any(np.isnan(X_train))

False

In [0]:
# Confirm there is no nan in  test dataset
np.any(np.isnan(X_test))

False

###**Feature scaling**

Feature scaling is a method used to normalize the range of independent variables or features of data. In data processing, it is also known as data normalization.

We need to normalize our independent variables. We use robust scaler to do this.

In [0]:
cols = X_train.columns


scaler = RobustScaler()

X_train = scaler.fit_transform(X_train)

X_test = scaler.transform(X_test)
X_train = pd.DataFrame(X_train, columns=[cols])
X_test = pd.DataFrame(X_test, columns=[cols])

##**Random Forest Classifier model with default parameters- 10 decision trees**

In [0]:

# Intiate the randomforestclassifier 
rf = RandomForestClassifier(random_state=0)

# fit the model
rf.fit(X_train, y_train)

# Predict the Test set results
y_pred = rf.predict(X_test)

# Check accuracy score 

from sklearn.metrics import accuracy_score

print('Model accuracy score with 10 decision-trees : {0:0.4f}'. format(accuracy_score(y_test, y_pred)))

Model accuracy score with 10 decision-trees : 0.9779


In [0]:
#Check the error rate of the model.

print('Root Mean Squared Error:', np.sqrt(metrics.mean_squared_error(y_test, y_pred)))

Root Mean Squared Error: 0.14871752967788066


In [0]:
#Compute the confusion matrix to understand the actual versus predicted variables.
from sklearn.metrics import confusion_matrix
confusion = confusion_matrix(y_test,y_pred)
confusion


array([[ 17,   9],
       [  5, 602]])




*   **Findings**

*   Model accuracy was 98% with an error rate of 0.14. The model predicted 602 negative and 17 hypothyroid correctly
*   To improve the model perfomance, we increase the number of decision trees to 100, increase the max depth and reduce sample split to 20.



##**Random forest classifier using 100 decision trees**

In [0]:
# Run the classifier with n_estimators = 100

rf1 = RandomForestClassifier(n_estimators=100, random_state=0,max_depth=5, min_samples_split = 20)

# fit the model to the training set
rf1.fit(X_train, y_train)

RandomForestClassifier(bootstrap=True, ccp_alpha=0.0, class_weight=None,
                       criterion='gini', max_depth=5, max_features='auto',
                       max_leaf_nodes=None, max_samples=None,
                       min_impurity_decrease=0.0, min_impurity_split=None,
                       min_samples_leaf=1, min_samples_split=20,
                       min_weight_fraction_leaf=0.0, n_estimators=100,
                       n_jobs=None, oob_score=False, random_state=0, verbose=0,
                       warm_start=False)

In [0]:
# Predict on the test set results
y_pred1 = rf1.predict(X_test)


In [0]:
# Create a comparison frame between the actual and predicted target variable
comparison_frame = pd.DataFrame({'Actual': y_test.flatten(), 'Predicted': y_pred.flatten()})

comparison_frame.describe()

Actual   Predicted
count  633.000000  633.000000
mean     0.958926    0.965245
std      0.198619    0.183304
min      0.000000    0.000000
25%      1.000000    1.000000
50%      1.000000    1.000000
75%      1.000000    1.000000
max      1.000000    1.000000

In [0]:
# Check accuracy score 
print('Model accuracy score with 100 decision-trees : {0:0.4f}'. format(accuracy_score(y_test, y_pred_100)))

Model accuracy score with 100 decision-trees : 0.9779


In [0]:
#Check the error rate using root mean squared error

from sklearn import metrics
print('Root Mean Squared Error:', np.sqrt(metrics.mean_squared_error(y_test, y_pred)))

Root Mean Squared Error: 0.14871752967788066


Error rate is very low. 0.14.Shows the model is relevantly good.

In [0]:
# Calculate a confusion matrix to identify what patients were predicted to be negative or have hypothyroid
confusion = confusion_matrix(y_test,y_pred1)
confusion

array([[ 17,   9],
       [  5, 602]])





* **Findings**


*   The model accuracy still remained the same 97% even after the number of decision trees.The error rate is very minimal : 0.14

*   The model predicted 602 patients were negative while in the actual they wrere negative.
*   The model predicted 17 patients had hypotyroid while in the actual they were.


*   This means the accuracy of the model is not affected by the change in decision trees.

*   However, we can still improve on the model using gradient boosting and see how it will perfom.






##**Gradient Boosting Classifier**

In [0]:
#Intiate the gradient boosting classifier
gradient = GradientBoostingClassifier(learning_rate =0.1,n_estimators=100,max_depth=3,min_samples_split=2) # defining my classifier as  gradient

#fit the train dataset in the classifier
gradient.fit(X_train,y_train)

GradientBoostingClassifier(ccp_alpha=0.0, criterion='friedman_mse', init=None,
                           learning_rate=0.1, loss='deviance', max_depth=3,
                           max_features=None, max_leaf_nodes=None,
                           min_impurity_decrease=0.0, min_impurity_split=None,
                           min_samples_leaf=1, min_samples_split=2,
                           min_weight_fraction_leaf=0.0, n_estimators=100,
                           n_iter_no_change=None, presort='deprecated',
                           random_state=None, subsample=1.0, tol=0.0001,
                           validation_fraction=0.1, verbose=0,
                           warm_start=False)

In [0]:
#Making a prediction
y_pred_g = gradient.predict(X_test)
y_pred_g

array([1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 0, 1, 1, 1, 1, 1, 1, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0, 1,
       1, 1, 1, 1, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0, 1, 1, 1, 1, 0, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 0, 1, 1, 1, 1, 1, 1, 1,

In [0]:
#Check the accuracy score of the gradient model
print("gradient_Accuracy score is :",metrics.accuracy_score(y_test, y_pred_g))

gradient_Accuracy score is : 0.9842022116903634


In [0]:
# Calculate a confusion matrix to identify what patients were predicted to be negative or have hypothyroid
confusion = confusion_matrix(y_test,y_pred_g)
confusion

array([[ 23,   3],
       [  7, 600]])



*   **Findings**
*   The accuracy score of the model increased to 98%.

*   The model predicted 600 patients were negative while in the actual they wrere negative.
*   The model predicted 23 patients had hypothyroid while in the actual they were.
*   In this case, we can use the gradient boosting classifier model compared to random forest as it's perfomance increased .





##**SVM(Support Vector Machine)**

In [0]:
#For this, we will create svm before parameter tuning and after parameter tuning using rbf. This is because we are solving a classifier.
# SVM before parameter tuning
#svm  = SVC(kernel = 'linear',C=1.0,gamma='auto',random_state=2)

#SVM after parameter tuning. RBf is used to increase the dimension
clf = SVC(kernel = 'sigmoid',C=1.0,gamma='auto',random_state=0)


In [0]:
#  fitting the train into the model
#svm.fit(X_train,y_train)

#svm_1.fit(X_train,y_train)
clf.fit(X_train,y_train)

SVC(C=1.0, break_ties=False, cache_size=200, class_weight=None, coef0=0.0,
    decision_function_shape='ovr', degree=3, gamma='auto', kernel='sigmoid',
    max_iter=-1, probability=False, random_state=0, shrinking=True, tol=0.001,
    verbose=False)

In [0]:
# Now that we have trained our model, let's test how well it can predict if a patient is negattive or positive for hypothyroid
#Making predictions
#y_pred_svc = svm.predict(X_test)

#Making predictions with parameter tuning
y_pred1 = clf.predict(X_test)

In [0]:
#Check accuracy of model before setting any parameters
print("Accuracy with linear kernel:",metrics.accuracy_score(y_test, y_pred_svc))

#Accuracy score using sigmoid function
print(accuracy_score(y_test,y_pred))

Accuracy with linear kernel: 0.976303317535545
0.9652448657187994





*   **Findings**


*    Reason behind not plotting the graph to show the hyperplane  is that we were dealing with multiple features so it would have been pretty hard to visualize.

*   The accuracy score of models was 97.6% after tuning the parameters using sigmoid the model accuracy dropped to 96%.





##**Further questions**


*   Based on the fact that our dataset had many patients with negative results on hypothyroid,did it affect the model accuracy to be that high even after using advanced models?






##**Challenges**


*   Despite using different classification models, differences between the model accuracy was very minimal.


*   It was a challenge picking the best model for prediction since they all had almost the same accuracy score and very low error rate.






##**Future prospects**


*   Dataset should be evenly distributed interms of status to avoid model overfitting.

